### ISSUE TITLE
False positive on `UnnecessaryLet`

### ISSUE THREAD

**Issue by atulgpt** [detekt/detekt#6947]
## Expected Behavior
```kotlin
@Test
fun `does not report let with which result in nested safe call with invoke`() {
    val findings = subject.compileAndLintWithContext(
        env,
        """
            class ClickContainer(val onClick: () -> Unit)

            fun test(callbacks: List<ClickContainer>) {
                callbacks
                    .firstOrNull()?.let {
                        it.onClick()
                    }
                callbacks
                    .firstOrNull()?.let {
                        it.onClick.invoke()
                    }
            }
        """.trimIndent()
    )
    assertThat(findings).isEmpty()
}
```
The above TC should pass

## Observed Behavior
The above TC is failing

## Context
In the above TC, in the below code section
```kotlin
callbacks
    .firstOrNull()?.let {
        it.onClick()
    }
```
we can remove `let` but the resulting code(`callbacks?.firstOrNull()?.onClick?.invoke()`) will not look good and result in other violations of cascading safe call

As a workaround, we can use `run` which doesn't trigger the rule. But the mentioned case should also pass

**Comment by BraisGabin** [comment]
I'm not 100% sure of this change. What do @detekt/maintainers think?

I agree that `callbacks?.firstOrNull()?.onClick?.invoke()` is not ideal. But `let` adds complexity to read too. But it adds more or less the same complexity as `run` does. Probably even less because it doesn't shadow `this`. So I'm not sure what we should do in this case.

**Comment by TWiStErRob** [comment]
I think `run` is not a workaround, but **a** solution here. Yes it's the same-ish complexity as `let`, but that's just syntax, there's also semantic meaning behind `let`/`also`/`run`/`apply`/`with`, that's why there's so many with different signatures. Sadly [the docs](https://kotlinlang.org/docs/scope-functions.html) focus very heavily (dare I say, only) on syntax.

In the above example you "select the first callback to invoke", i.e. `callbacks?.firstOrNull()?.run { onClick() }` a valid expression of that requirement, you get an object and then call operations on it.

<details><summary>Note: to fix the chain call detection (which is totally reasonable in this complexity)</summary>

Extract a val or fun that does a part of the expression, named!:
```kotlin
private val firstCallback: CB?
    get() = callbacks?.firstOrNull() // or callbacks.orEmpty().firstOrNull()

firstCallback?.onClick?.invoke()
```
or
```kotlin
internal fun CB.onClick() {
   this.onClick.invoke()
}
callbacks?.firstOrNull()?.onClick()
```
or even together
```kotlin
firstCallback?.onClick()
```

To notice these possibilities, I recommend you play some object calisthenics exercises, they can be really fun.

</details>

**Comment by BraisGabin** [comment]
So, which is your position about this PR? Don't merge and use `run`?

**Comment by atulgpt** [comment]
Hi, @TWiStErRob to me `let` seems more reasonable here. I generally use `let` to invoke some method on a nullable value when it is not null. (TBH I don't use `run` much 🤷🏼‍♂️)

Also when you see the TCs of `UnnecessaryLet` most of the TCs enforce patterns like `a?.someFun()` instead of `a?.let { it.someFun() }` which is not possible in this case. As in reality `it.onClick()` is actually `it.onClick.invoke()`

I also found one similar TC which somewhat does the same thing around `invoke` which fixed https://github.com/detekt/detekt/issues/5236
```kotlin
@Test
fun `does not report lets with invoke operator calls`() {
    val findings = subject.compileAndLintWithContext(
        env,
        """
            fun f(callback: ((Int) -> Int)?) {
                callback?.let { that ->
                    that(42)
                }
            }
        """.trimIndent()
    )
    assertThat(findings).isEmpty()
}
```

**Comment by fluxxion82** [comment]
I think I'm seeing the same thing. Somehow I suddenly started getting a message about an UnnecessaryLet. I cherry-picked a commit to a different branch and for some reason the code style check started complaining about this. The code seems fine and doesn't get triggered on other branches. When I try to update the baseline it's not getting added either. I'm using the same versions of gradle, kotlin, plugins between branches, so I'm very confused about how I'm seeing this.

**Comment by 3flex** [comment]
Kotlin 2.0.0-Beta4 seems to solve the original case, so I don't think we need any behaviour change since the compiler will allow improved syntax to be used (without `let`): https://kotlinlang.org/docs/whatsnew-eap.html#properties-with-function-types

This compiles with the latest beta:
```kotlin
callbacks.firstOrNull()?.onClick()
```
